### DISCUSSION TITLE
Enforce Virtual Display Primary

### DISCUSSION THREAD

**Discussion by simon3948** [ClassicOldSong/Apollo#169] [General]
Hello ClassicOldSong,

I have stumbled across this fork today out of complete luck, wanted to say im very impressed with the seamlessness of the virtual display functionality, it baffles me how this hasnt been added to the main Sunshine/moonlight stack. Very interested to see where this project goes.

There is an option for "Enforce Virtual Display Primary" option, I believe this forces all apps to be enabled with a virtual display by default with indicated caveat: Virtual display will always be set to primary when client requests to use virtual display. 

Would it be possible to configure virtual display to not take as primary on a per client basis/ per apollo instance?

I am interested in using this stack for using an old android tablet as an extra monitor so the virtual display taking over the main monitor is not what I want. Apollo/artemis is like 90% there for my use case. Would also be great to have an auto reconnect function on the artemis side when the PC restarts.

I tried amending the windows primary display back to my main monitor and disconnecting/reconnecting Atemis client but it still takes as primary.

Do you think this is something you would be interested in adding in the future?
 I see in artemis there is an option for "External Display Mode (Beta)" Does this do anything? I couldn't find any information about it.

Again, thanks for the great work!

**Comment by ClassicOldSong** [comment]
If you explicitly start with "Launch in Virtual Display" or have "Use Virtual Display" checked in Artemis it'll always set to primary. It's a safety feature to (mostly) ensure you can see the app you launched on your client.

The solution to this is, check "Always use Virtual Display" and "Use app's own identity" in the specific app you're launching in Apollo. In this case Apollo won't force the virtual display to be primary or not, it'll completely depend on your own configuration in Windows settings.

Or another solution: upgrade to Windows11 24H2 as it broke the functionality for setting primary 🤣

**Reply by ClassicOldSong** [reply]
The External Display Mode is for connecting an external monitor to your android device, like a portable display or AR glasses.

**Reply by simon3948** [reply]
Thats exactly what I'm looking for thanks!
Is there anyway to set artemis to go into a suspended state (Screenoff/black screen) when the host PC turns off/restarts and continuously retry to reconect?
I would like to have this display permentantly connected and reconnect automatically on pc startup if possible.

**Reply by ClassicOldSong** [reply]
Well no, but you might be able to do it with some automation tools. Here is a tutorial containing instructions to auto connect: https://github.com/drajabr/My-Sunshine-setup

**Reply by mblennegard** [reply]
> If you explicitly start with "Launch in Virtual Display" or have "Use Virtual Display" checked in Artemis it'll always set to primary. It's a safety feature to (mostly) ensure you can see the app you launched on your client.
> 
> The solution to this is, check "Always use Virtual Display" and "Use app's own identity" in the specific app you're launching in Apollo. In this case Apollo won't force the virtual display to be primary or not, it'll completely depend on your own configuration in Windows settings.
> 
> Or another solution: upgrade to Windows11 24H2 as it broke the functionality for setting primary 🤣

What exact settings are supposed to be used to have the virtual display be set as Primary on connect, because for me all settings I have tried simply displays the secondary screen (the virtual display) on the client device.

Also, is it possible to turn off the physical screen and only have the currently connected client virtual display be "turned on"?

**Reply by ClassicOldSong** [reply]
Use Windows's native display settings. You only need to set the display configuration once, and they'll remembered.
The automatic primary settings was removed in later updates as Windows native method is much more stable.

**Reply by mblennegard** [reply]
> Use Windows's native display settings. You only need to set the display configuration once, and they'll remembered. The automatic primary settings was removed in later updates as Windows native method is much more stable.

Whoa, that was fast reply. 👍
I tried using the Windows native display settings as well, setting the virtual display adapter as primary there, thinking the virtual display adapter would be removed and therefore my physical display reverting back to primary upon disconnecting the client but that only made it so I must manually set the physical screen back to primary before disconnecting.

What exact settings should be used within Windows?
Also, what you are referring to, is that only for making the virtual display adapter Primary upon connect or turning off the physical display as well?

**Reply by ClassicOldSong** [reply]
You need to terminate the stream. It'll remove the virtual display and everything else will be reverted automatically.

If you want to turn off your physical display, you need to set the display mode to "extended" during stream.

**Reply by mblennegard** [reply]
> You need to terminate the stream. It'll remove the virtual display and everything else will be reverted automatically.
> 
> If you want to turn off your physical display, you need to set the display mode to "extended" during stream.

Ah, I thought the hotkey Start + Select + L1 + R1 was to terminate the session, but I noticed the wording now when connecting it says _disconnect_. Closing the application after disconnecting seems to terminate the entire session.
This solves the "primary display" part.

However, turning the physical display off I cannot figure out.
In case I explained it poorly before, I am not talking about physically turning the display off manually but rather disabling the display automatically whilst streaming so the display stays off/turns off automatically.

**Reply by ClassicOldSong** [reply]
Yes, when you're in stream with the virtual display active, either press Win+P and select External display only or go to Windows settings select "Show only on 2"

Windows does a very bad job here using different wordings for the same thing.

**Reply by mblennegard** [reply]
> Yes, when you're in stream with the virtual display active, either press Win+P and select External display only or go to Windows settings select "Show only on 2"
> 
> Windows does a very bad job here using different wordings for the same thing.

In case anyone else finds this discussion thread in the future, setting the "Show only on 2" etc. in Windows whilst streaming actually solves the "issue" I was having, as the root cause for both my issues were due to not terminating the connection, only disconnecting.
